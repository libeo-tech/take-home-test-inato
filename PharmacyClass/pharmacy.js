export class Pharmacy {
  constructor(drugs = []) {
    this.drugs = drugs;
  }
  updateBenefitValue() {
    for (var i = 0; i < this.drugs.length; i++) {
      if (this.drugs[i].name !== "Magic Pill") {
        this.drugs[i].updateExpiresIn();
        if (this.drugs[i].benefit < 50 && this.drugs[i].benefit > 0)
          this.drugs[i].updateBenefit();
      }
    }
    return this.drugs;
  }
}
