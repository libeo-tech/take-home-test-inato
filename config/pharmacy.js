export const pharmacyConfig = {
  maxBenefit: 50,
  minBenefit: 0,
};
