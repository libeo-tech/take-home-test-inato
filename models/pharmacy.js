export class Pharmacy {
  constructor(drugs = []) {
    this.drugs = drugs;
  }
  updateBenefitValue() {
    this.drugs.forEach(drug => drug.updateBenefitValue());

    return this.drugs;
  }
}
