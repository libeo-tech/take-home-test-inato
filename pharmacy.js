export class Pharmacy {
  constructor(drugs = []) {
    this.drugs = drugs;
  }
  updateBenefitValue() {
    return this.drugs.map((drug) => drug.updateBenefit());
  }
}
