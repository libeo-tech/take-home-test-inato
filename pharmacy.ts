interface DrugData {
  readonly name: string;
  expiresIn: number;
  benefit: number;
}
export class Drug implements DrugData {
  constructor(
    public readonly name: string,
    public expiresIn: number,
    public benefit: number,
  ) {
  }

  public toJSON = (): DrugData => ({
    name: this.name,
    expiresIn: this.expiresIn,
    benefit: this.benefit
  });

  public toObject = (): DrugData => ({
    name: this.name,
    expiresIn: this.expiresIn,
    benefit: this.benefit
  });

  public benefitRevaluate(): void {
    this._benefitRevaluate();
    this.updateExpires();
  }

  protected _benefitRevaluate(): void {
    // TODO: Missing logic
  }

  protected updateExpires(): void {
    this.expiresIn = this.expiresIn - 1;
  }
}
export class Pharmacy {
  constructor(public drugs = []) {
    this.drugs = drugs;
  }
  updateBenefitValue() {
    for (var i = 0; i < this.drugs.length; i++) {
      if (
        this.drugs[i].name != "Herbal Tea" &&
        this.drugs[i].name != "Fervex"
      ) {
        if (this.drugs[i].benefit > 0) {
          if (this.drugs[i].name != "Magic Pill") {
            this.drugs[i].benefit = this.drugs[i].benefit - 1;
          }
        }
      } else {
        if (this.drugs[i].benefit < 50) {
          this.drugs[i].benefit = this.drugs[i].benefit + 1;
          if (this.drugs[i].name == "Fervex") {
            if (this.drugs[i].expiresIn < 11) {
              if (this.drugs[i].benefit < 50) {
                this.drugs[i].benefit = this.drugs[i].benefit + 1;
              }
            }
            if (this.drugs[i].expiresIn < 6) {
              if (this.drugs[i].benefit < 50) {
                this.drugs[i].benefit = this.drugs[i].benefit + 1;
              }
            }
          }
        }
      }
      if (this.drugs[i].name != "Magic Pill") {
        this.drugs[i].expiresIn = this.drugs[i].expiresIn - 1;
      }
      if (this.drugs[i].expiresIn < 0) {
        if (this.drugs[i].name != "Herbal Tea") {
          if (this.drugs[i].name != "Fervex") {
            if (this.drugs[i].benefit > 0) {
              if (this.drugs[i].name != "Magic Pill") {
                this.drugs[i].benefit = this.drugs[i].benefit - 1;
              }
            }
          } else {
            this.drugs[i].benefit =
              this.drugs[i].benefit - this.drugs[i].benefit;
          }
        } else {
          if (this.drugs[i].benefit < 50) {
            this.drugs[i].benefit = this.drugs[i].benefit + 1;
          }
        }
      }
    }

    return this.drugs;
  }
}
