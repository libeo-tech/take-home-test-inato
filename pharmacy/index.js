import {
  updateHerbalTea,
  updateFervex,
  updateDefaultDrug,
  updateDafalgan
} from "./drugs";

const updateDrugs = drugs => {
  const drugsUpdated = drugs.map(drug => {
    let drugUpdated;
    switch (drug.name) {
      case "Herbal Tea":
        drugUpdated = updateHerbalTea(drug);
        break;
      case "Fervex":
        drugUpdated = updateFervex(drug);
        break;
      case "Magic Pill":
        drugUpdated = { ...drug };
        break;
      case "Dafalgan":
        drugUpdated = updateDafalgan(drug);
        break;
      case "Doliprane":
      default:
        drugUpdated = updateDefaultDrug(drug);
        break;
    }
    return drugUpdated;
  });
  return drugsUpdated;
};

export const createPharmacy = (drugs = []) => {
  const update = () => {
    drugs = updateDrugs(drugs);
    return drugs;
  };

  return {
    drugs,
    update
  };
};
