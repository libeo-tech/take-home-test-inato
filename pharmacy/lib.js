import { MIN_BENEFIT_VALUE, MAX_BENEFIT_VALUE } from "./constants";

export const limitBenefitToMin = drug => {
  if (drug.benefit < MIN_BENEFIT_VALUE) {
    return { benefit: MIN_BENEFIT_VALUE };
  }
  return drug;
};

export const limitBenefitToMax = drug => {
  if (drug.benefit > MAX_BENEFIT_VALUE) {
    return { benefit: MAX_BENEFIT_VALUE };
  }
  return drug;
};

export const decreaseExpiresIn = drug => {
  return {
    expiresIn: drug.expiresIn - 1
  };
};
