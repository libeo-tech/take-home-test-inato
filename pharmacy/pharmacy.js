export class Pharmacy {
  constructor(drugs = []) {
    this.drugs = drugs;
  }
  updateBenefitValue() {
    this.drugs.forEach(drug => {
      drug.aging();
    });

    return this.drugs;
  }
}
