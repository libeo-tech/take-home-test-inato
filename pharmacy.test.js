import { Drug, Pharmacy } from "./pharmacy";

describe("Pharmacy", () => {
  it("should decrease the benefit and expiresIn", () => {
    expect(new Pharmacy([new Drug("test", 2, 3)]).updateBenefitValue()).toEqual(
      [new Drug("test", 1, 2)]
    );
  });
  it("should decrease the benefit twice and expiresIn", () => {
    expect(new Pharmacy([new Drug("test", 0, 3)]).updateBenefitValue()).toEqual(
      [new Drug("test", -1, 1)]
    );
  });

  it("for Herbal Tea, should increase the benefit and decrease expiresIn", () => {
    expect(new Pharmacy([new Drug("Herbal Tea", 2, 3)]).updateBenefitValue()).toEqual(
      [new Drug("Herbal Tea", 1, 4)]
    );
  });
  it("for Herbal Tea, should increase the benefit twice and decrease expiresIn", () => {
    expect(new Pharmacy([new Drug("Herbal Tea", 0, 3)]).updateBenefitValue()).toEqual(
      [new Drug("Herbal Tea", -1, 5)]
    );
  });

  it("for Magic Pill, should do nothing", () => {
    expect(new Pharmacy([new Drug("Magic Pill", 2, 3)]).updateBenefitValue()).toEqual(
      [new Drug("Magic Pill", 2, 3)]
    );
  });

  it("for Fervex, should increase the benefit and decrease expiresIn", () => {
    expect(new Pharmacy([new Drug("Fervex", 20, 3)]).updateBenefitValue()).toEqual(
      [new Drug("Fervex", 19, 4)]
    );
  });
  it("for Fervex, should increase the benefit twice and decrease expiresIn", () => {
    expect(new Pharmacy([new Drug("Fervex", 7, 3)]).updateBenefitValue()).toEqual(
      [new Drug("Fervex", 6, 5)]
    );
  });
  it("for Fervex, should increase the benefit by three and decrease expiresIn", () => {
    expect(new Pharmacy([new Drug("Fervex", 2, 3)]).updateBenefitValue()).toEqual(
      [new Drug("Fervex", 1, 6)]
    );
  });

  it("for Dafalgan, should decrease the benefit and expiresIn", () => {
    expect(new Pharmacy([new Drug("Dafalgan", 2, 3)]).updateBenefitValue()).toEqual(
      [new Drug("Dafalgan", 1, 1)]
    );
  });
  it("for Dafalgan, should decrease the benefit twice and expiresIn", () => {
    expect(new Pharmacy([new Drug("Dafalgan", 0, 7)]).updateBenefitValue()).toEqual(
      [new Drug("Dafalgan", -1, 3)]
    );
  });
});
