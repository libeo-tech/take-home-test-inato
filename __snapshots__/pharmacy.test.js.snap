// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pharmacy Fervex [unit] should increases in benefit as its expiration date approaches 1`] = `
Array [
  Drug {
    "benefit": 11,
    "expiresIn": 19,
    "name": "Fervex",
  },
]
`;

exports[`Pharmacy Fervex [unit] should increases in benefit by 2 when <= 10 days 1`] = `
Array [
  Drug {
    "benefit": 13,
    "expiresIn": 8,
    "name": "Fervex",
  },
]
`;

exports[`Pharmacy Fervex [unit] should increases in benefit by 3 when <= 5 days 1`] = `
Array [
  Drug {
    "benefit": 16,
    "expiresIn": 3,
    "name": "Fervex",
  },
]
`;

exports[`Pharmacy General drug [unit] should decrease benefit & expiresIn 1`] = `
Array [
  Drug {
    "benefit": 2,
    "expiresIn": 1,
    "name": "test",
  },
]
`;

exports[`Pharmacy General drug [unit] should degrades benefit twice after expiration 1`] = `
Array [
  Drug {
    "benefit": 0,
    "expiresIn": -1,
    "name": "test",
  },
]
`;

exports[`Pharmacy General drug [unit] should not have benefit > 50 1`] = `
Array [
  Drug {
    "benefit": 50,
    "expiresIn": -19,
    "name": "Herbal Tea",
  },
]
`;

exports[`Pharmacy General drug [unit] should not have negative benefit 1`] = `
Array [
  Drug {
    "benefit": 0,
    "expiresIn": -2,
    "name": "test",
  },
]
`;

exports[`Pharmacy Herbal Tea [unit] should increases in benefit as its expiration date approaches 1`] = `
Array [
  Drug {
    "benefit": 6,
    "expiresIn": 4,
    "name": "Herbal Tea",
  },
]
`;

exports[`Pharmacy Herbal Tea [unit] should increases in benefit twice after the expiration date 1`] = `
Array [
  Drug {
    "benefit": 5,
    "expiresIn": -1,
    "name": "Herbal Tea",
  },
]
`;

exports[`Pharmacy Magic Pill [unit] should never expires nor decreases in benefit 1`] = `
Array [
  Drug {
    "benefit": 10,
    "expiresIn": 1,
    "name": "Magic Pill",
  },
]
`;
